graph/appender/response_time.go

```diff
@@ -48,60 +50,98 @@ func (a ResponseTimeAppender) appendGraph(trafficMap graph.TrafficMap, namespace
 		int(a.Duration.Seconds()), // range duration for the query
-		"le,source_service,source_version,destination_service,destination_version")
+		groupBy)
 	inVector := promQuery(query, time.Unix(a.QueryTime, 0), client.API())
 
 	// create map to quickly look up responseTime
 	responseTimeMap := make(map[string]float64)
-	populateResponseTimeMap(responseTimeMap, &outVector)
-	populateResponseTimeMap(responseTimeMap, &inVector)
+	a.populateResponseTimeMap(responseTimeMap, &unkVector)
+	a.populateResponseTimeMap(responseTimeMap, &outVector)
+	a.populateResponseTimeMap(responseTimeMap, &inVector)
 
 	applyResponseTime(trafficMap, responseTimeMap)
 }
 
 func applyResponseTime(trafficMap graph.TrafficMap, responseTimeMap map[string]float64) {
-	// TODO FIX Name --> Workload to compile
 	for _, s := range trafficMap {
 		for _, e := range s.Edges {
-			key := fmt.Sprintf("%s %s %s %s", e.Source.Workload, e.Source.Version, e.Dest.Workload, e.Dest.Version)
+			key := fmt.Sprintf("%s %s", e.Source.ID, e.Dest.ID)
 			e.Metadata["responseTime"] = responseTimeMap[key]
 		}
 	}
 }
 
-func populateResponseTimeMap(responseTimeMap map[string]float64, vector *model.Vector) {
+func (a ResponseTimeAppender) populateResponseTimeMap(responseTimeMap map[string]float64, vector *model.Vector) {
+	//groupBy := "le,source_workload_namespace,source_workload,source_app,source_version,destination_service_namespace,destination_workload,destination_app,destination_version"
 	for _, s := range *vector {
 		m := s.Metric
-		sourceSvc, sourceSvcOk := m["source_service"]
+		sourceWlNs, sourceWlNsOk := m["source_workload_namespace"]
+		sourceWl, sourceWlOk := m["source_workload"]
+		sourceApp, sourceAppOk := m["source_app"]
 		sourceVer, sourceVerOk := m["source_version"]
-		destSvc, destSvcOk := m["destination_service"]
+		destSvcNs, destSvcNsOk := m["destination_service_namespace"]
+		destSvcName, destSvcNameOk := m["destination_service_namespace"]
+		destWl, destWlOk := m["destination_workload"]
+		destApp, destAppOk := m["destination_app"]
 		destVer, destVerOk := m["destination_version"]
-		if !sourceSvcOk || !sourceVerOk || !destSvcOk || !destVerOk {
+		if !sourceWlNsOk || !sourceWlOk || !sourceAppOk || !sourceVerOk || !destSvcNsOk || !destSvcNameOk || !destWlOk || !destAppOk || !destVerOk {
 			log.Warningf("Skipping %v, missing expected labels", m.String())
 			continue
 		}
 
-		key := fmt.Sprintf("%s %s %s %s", sourceSvc, sourceVer, destSvc, destVer)
+		// For source-side failures (Fault injection, open circuit breakers, etc) the
+		// destination_workload will be unknown (because the request never made it to the
+		// destination proxy.  But we still want to record the request in some
+		// way. So, depending on the graph type and labeling, assign the destination
+		// service name (i.e. the requested service) to the most applicable destination field.
+		// note: this code is duplicated in graph.go for any changes/fixes
+		if string(destApp) == graph.UnknownApp {
```

Comment:
the comment says "...the destination_workload will be unknown..." but yet the if-stmt checks if destApp is unknown. Shouldn't this be checking if "destWl" is unknown?

Reply:
The power of comments.  Good catch.  DestApp will also be unknown, but that could also happen in app labels were not being used.  So for our current use cases this works as is, but it does need to change to cover all cases. I'll make the update.